### DISCUSSION TITLE
Compiler frontend

### DISCUSSION THREAD

**Discussion by DosWorld** [open-watcom/open-watcom-v2#1006] [Ideas]
Hello!

I am try to play with compilers and want start try create one more frontend for watcom and merge it to V2.
Of course, this will not happen soon, probably in a year(s).

What you think about it? Do you have any wishes that should be considered?

**Comment by jmalak** [comment]
What purpose of such front-end?

**Comment by DosWorld** [comment]
I just want start working and add language from N. Wirth languages-family.
So, I am, run forward, and ask what can be at end of road.

**Reply by jmalak** [reply]
May be LLVM is more suitable for it because it is more optimized.
OW code generator is much older.

**Comment by pchapin** [comment]
I have a strong interest in Ada, which has syntax in the Pascal family, but unlike some languages it is still actively used (particularly in the embedded systems space). I have fantasized about a project to create an Ada front end for the Open Watcom code generator and tool set. I can rationalize such a project in several ways: since Open Watcom targets some lesser used systems, this would bring Ada to those systems. Also, since mixed Ada/C programs are common, interaction with the Open Watcom C compiler would be interesting. For example, the Ada runtime library could be implemented on top of the Open Watcom C runtime library.

It's a fantasy because Ada is an enormous language on a scale like (or even exceeding?) that of C++. A full Ada compiler front end is a huge undertaking. Of course, one could start with a "mini" version of the language. Ada's syntax is much more straightforward than C++'s syntax. Ada's complexity is mostly semantic. However, a mini-Ada could eliminate much of that.

Plus, the tentative name for a Watcom Ada Compiler might be `wac` and who can pass up working on a tool with a name like that? :)

**Comment by DosWorld** [comment]
My hobby is programming for ms-dos, so i paid attention to Watcom. I known about watcom code generation and about old gnu toolchain (djgpp).
Okay, as far as i understand your answers - you're open to discussion (when+if something is done).
That's enough for me.

I wouldn't start with Ada, I'd choose something simpler (and familiar to me) project. (mini-frontend sounds ugly, like a "student's compilers". i dont want bring trash to watcom) OK. Something needs to be done first. And see.

**Reply by jmalak** [reply]
If you concentrate to DOS platfom then OW code generator is suitable for it.
OW code generator is one which support well 16-bit segmented Intel architecture.
It is not a problem to support you simply raise issue on github.
